### PR TITLE
[READY] Add tests for cache completion

### LINKED ISSUE #430: Strange caching behavior on completions requests
Hi there,

In testing YCMD as a remote completion API, I appear to be hitting a weird caching bug. I'm just testing against Jedi autocompletion currently.

For the buffer

```python
import requests
import datetime

requests.
```

I (on the remote server where ycmd is running), write out the buffer to the filesystem, send a `FileReadyToParse` event to YCMD, and then ask for completions. This works okay, I get the appropriate results for `requests`.

However, If I change the buffer to:

```python
import requests
import datetime

datetime.
```

and perform the same actions, I still get returned the completion results for `requests` and not `datetime`. I believe this to be because of YCMD's caching on line and character offsets. Is there an API for saying "this buffer needs to be reparsed"? I couldn't find one in the sample client.

### PR DESCRIPTION
**Pull request by micbou**
Add tests for cache completion and clean up CompletionsCache class.

---
This change is [Reviewable](https://reviewable.io/reviews/valloric/ycmd/442)

### CODE REVIEW

File: ycmd/tests/get_completions_test.py [hunk 1]
```diff
@@ -199,3 +199,191 @@ def GetCompletions_SemanticCompleter_WorksWhenTriggerIsIdentifier_test(
...
+                                    column_num = 12 )
+
+    results = app.post_json( '/completions',
+                             completion_data ).json[ 'completions' ]
+    assert_that(
+      results,
+      has_items( CompletionEntryMatcher( 'attributeA' ) )
+    )
+
+    # Start column is 8
```
Comment (vincentwoo) [inline ycmd/tests/get_completions_test.py]: the comments say start columns 8 and 9, respectively, but these calls both appear to be using column 12. am I misreading something?

Comment (Valloric) [inline ycmd/tests/get_completions_test.py]: @vincentwoo Try to leave line-level comments in Reviewable.io, not on Github. We use Reviewable for reviews and it doesn't work well with Github line-level comments. :)

**Comment by Valloric** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/request_wrap.py, line 41 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/442#-KDkIsy6OUI-uKYXkoby:-KDkIsy6OUI-uKYXkobz:316403114) ([raw file](https://github.com/valloric/ycmd/blob/afb8e95fc5efbeaab28b93254c5f01318341d8ad/ycmd/request_wrap.py#L41)):*
Maybe `line_value_until_start`?

---

*[ycmd/tests/get_completions_test.py, line 272 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/442#-KDkJEySFC1V4G17QL1S:-KDkJEySFC1V4G17QL1T:-1906144543) ([raw file](https://github.com/valloric/ycmd/blob/afb8e95fc5efbeaab28b93254c5f01318341d8ad/ycmd/tests/get_completions_test.py#L272)):*
Wording is a bit funky here; how about "We ask for candidates twices because of..."?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/442)*

**Comment by Valloric** [comment]
:lgtm: with minor comments. Thanks!

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/442#-:-KDkKXL2GjYcTWf5CB0q:459185884)*

**Comment by micbou** [comment]
Update the PR to only add tests and do some clean-up.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[ycmd/tests/get_completions_test.py, line 272 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/442#-KDkJEySFC1V4G17QL1S:-KDkdVxPAIa9MOs3cHeJ:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/afb8e95fc5efbeaab28b93254c5f01318341d8ad/ycmd/tests/get_completions_test.py#L272)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/442#-:-KDkdpxep5zAXknToNDD:-137903387)*

**Comment by Valloric** [comment]
Yay, more tests! :)

:lgtm: 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/442#-:-KDkeUpFnwuvPYd89hjr:276031376)*

**Comment by puremourning** [comment]
nice :lgtm: 

@homu r+

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/442#-:-KDl0HxM0Gc4vQPcIw1J:775043104)*

**Comment by puremourning** [comment]
Reviewed 2 of 3 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/442)*

**Pull request merged**